src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -27,11 +27,38 @@ public ChangedText(SourceText oldText, SourceText newText, ImmutableArray<TextCh
             RoslynDebug.Assert(oldText != null);
             Debug.Assert(oldText != newText);
             Debug.Assert(!changeRanges.IsDefault);
+            RequiresChangeRangesAreValid(oldText, newText, changeRanges);
 
             _newText = newText;
             _info = new ChangeInfo(changeRanges, new WeakReference<SourceText>(oldText), (oldText as ChangedText)?._info);
         }
 
+        private static void RequiresChangeRangesAreValid(
+            SourceText oldText, SourceText newText, ImmutableArray<TextChangeRange> changeRanges)
+        {
+            var deltaLength = 0;
+            foreach (var change in changeRanges)
+                deltaLength += change.NewLength - change.Span.Length;
+
+            if (oldText.Length + deltaLength != newText.Length)
+                throw new InvalidOperationException("Delta length difference of change ranges didn't match before/after text length.");
```

Comment:
💭 Seems like this would be `ArgumentException`

Reply:
These are internal checks of our own invariants.  i.e. by the time we got here, these shoudl apply.  I would prefer to just use Contract.ThrowIf... but id doesn't seem to be at this level.